mmseg/models/decode_heads/isnet_head.py

```diff
@@ -0,0 +1,349 @@
+                                                        0), preds_iter.permute(
+                                                            1, 0)
+            # (H*W, )
+            argmax = preds_iter.argmax(1)
+            for clsid in range(num_classes):
+                mask = (argmax == clsid)
+                if mask.sum() == 0:
+                    continue
+                feats_iter_cls = feats_iter[mask]
+                preds_iter_cls = preds_iter[:, clsid][mask]
+                # weight = F.softmax(preds_iter_cls, dim=0)
+                weight = torch.softmax(preds_iter_cls, dim=0)
+                feats_iter_cls = feats_iter_cls * weight.unsqueeze(-1)
+                feats_iter_cls = feats_iter_cls.sum(0)
+                feats_sl[batch_idx][mask] = feats_iter_cls
+        feats_sl = feats_sl.reshape(batch_size, h, w, num_channels)
+        feats_sl = feats_sl.permute(0, 3, 1, 2).contiguous()
+        feats_sl = self.correlate_net(inputs, feats_sl)
+        if hasattr(self, 'bottleneck'):
+            feats_sl = self.bottleneck(torch.cat([feats_il, feats_sl], dim=1))
+        return feats_sl
+
+
+@MODELS.register_module()
+class ISNetHead(BaseDecodeHead):
+
+    def __init__(self, transform_channels, concat_input, shortcut,
+                 dropout_ratio, **kwargs):
+        super().__init__(**kwargs)
+
+        ilc_cfg = {
+            'feats_channels': self.channels,
+            'transform_channels': transform_channels,
+            'concat_input': concat_input,
+            'norm_cfg': self.norm_cfg,
+            'act_cfg': self.act_cfg,
+            'align_corners': self.align_corners
+        }
+
+        slc_cfg = {
+            'feats_channels': self.channels,
+            'transform_channels': transform_channels,
+            'concat_input': concat_input,
+            'norm_cfg': self.norm_cfg,
+            'act_cfg': self.act_cfg,
+        }
+
+        self.in_channels = self.in_channels[-1]
+
+        self.bottleneck = ConvModule(
+            self.in_channels,
+            self.channels,
+            kernel_size=3,
+            stride=1,
+            padding=1,
+            conv_cfg=self.conv_cfg,
+            norm_cfg=self.norm_cfg,
+            act_cfg=self.act_cfg
+        )
+
+        self.ilc_net = ImageLevelContext(**ilc_cfg)
+        self.slc_net = SemanticLevelContext(**slc_cfg)
+
+
+        self.decoder_stage1 = nn.Sequential(
+            ConvModule(
+                self.channels,
+                self.channels,
+                kernel_size=1,
+                stride=1,
+                padding=0,
+                conv_cfg=self.conv_cfg,
+                norm_cfg=self.norm_cfg,
+                act_cfg=self.act_cfg),
+            nn.Dropout2d(dropout_ratio),
+            nn.Conv2d(
+                self.channels,
+                self.num_classes,
+                kernel_size=1,
+                stride=1,
+                padding=0,
+                bias=True),
+        )
+
+        if shortcut['is_on']:
```

Comment:
Could unpack the `shortcut` dict to several parameters, like `with_shortcut`, `shortcut_in_channels`, `shortcut_feat_channels`

If you prefer to use dict, should check the keys 'is_on', 'in_channels' and 'feats_channels' in the dict.